### ISSUE TITLE
[EarlyCSE] recognize commutative intrinsics with >2 args

### ISSUE THREAD

**Issue by rotateright** [llvm/llvm-project#46754]
|  |  |
| --- | --- |
| Bugzilla Link | [47410](https://llvm.org/bz47410) |
| Version | trunk |
| OS | All |
| CC | @RKSimon |

## Extended Description 
EarlyCSE was enhanced to recognize intrinsics that are commutative (https://reviews.llvm.org/D86798), but it is currently restricted to arguments with exactly 2 args. 

But there are a small set of intrinsics that have commutative first 2 operands with extra args after that:

declare i16 @&#8203;llvm.smul.fix.i16(i16, i16, i32)
define i16 @&#8203;intrinsic_3_args(i16 %x, i16 %y) {
  %m1 = call i16 @&#8203;llvm.smul.fix.i16(i16 %x, i16 %y, i32 1)
  %m2 = call i16 @&#8203;llvm.smul.fix.i16(i16 %y, i16 %x, i32 1)
  %r = sub i16 %m1, %m2
  ret i16 %r
}

$ opt commute.ll -S -gvn
define i16 @&#8203;intrinsic_3_args(i16 %x, i16 %y) {
  %m1 = call i16 @&#8203;llvm.smul.fix.i16(i16 %x, i16 %y, i32 1)
  ret i16 0
}

$ opt commute.ll -S -early-cse
define i16 @&#8203;intrinsic_3_args(i16 %x, i16 %y) {
  %m1 = call i16 @&#8203;llvm.smul.fix.i16(i16 %x, i16 %y, i32 1)
  %m2 = call i16 @&#8203;llvm.smul.fix.i16(i16 %y, i16 %x, i32 1)
  %r = sub i16 %m1, %m2
  ret i16 %r
}

**Comment by nikic** [comment]
Yes, this is already supported, see for example: https://github.com/llvm/llvm-project/blob/84f23eb3113f2e75d1a2e45db1b5c570a5d2f4c5/llvm/test/Transforms/EarlyCSE/commute.ll#L1002
